scripts/check.sh

```diff
@@ -0,0 +1,153 @@
+#!/bin/bash
+
+# Copyright 2019 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Checks that the current state of the tree is sane and optionally auto-fixes
+# errors automatically. Meant for interactive use.
+
+function usage() {
+  cat <<EOF
+USAGE: scripts/check.sh [--allow-dirty] [--commit] [<revision>]
+
+Runs auto-formatting scripts, source-tree checks, and linters on any files that
+have changed since master.
+
+By default, any changes are left as uncommited changes in the working tree. You
+can review them with git diff. Pass --commit to automatically commit any changes.
+
+Pass an alternate revision to use as the basis for checking changes.
+
+OPTIONS:
+
+  --allow-dirty
+    By default, check.sh requires a clean working tree to keep any generated
+    changes separate from logical changes.
+
+  --commit
+    Commit any auto-generated changes.
+
+  <revision>
+    Specifies a starting revision other than the default of master.
+
+
+EXAMPLES:
+
+  check.sh
+    Runs automated checks and formatters on all changed files since master.
+    Check for changes with git diff.
+
+  check.sh --commit
+    Runs automated checks and formatters on all changed files since master and
+    commits the results.
+
+  check.sh --commit HEAD
+    Runs automated checks and formatters on all changed files since the last
+    commit.
+
+  check.sh --allow-dirty HEAD
+    Runs automated checks and formatters on all changed files since the last
+    commit and intermingles the changes with any pending changes. Useful for
+    interactive use from an editor.
+
+EOF
+}
+
+set -euo pipefail
+unset CDPATH
+
+# Change to the top-directory of the working tree
+top_dir=$(git rev-parse --show-toplevel)
+cd "$top_dir"
```

Comment:
style guideline prefers (but doesn't require) brace-quoting, i.e. `cd "${top_dir}"`, (excepting single character and shell special variables i.e. `"$1"`). Here and throughout.